### PR TITLE
Track Down Unhandled rejections

### PR DESCRIPTION
**Pull request by acinader**
fixes: #5500 see #5496, #5488 

Add @stage88 0eff786 to fix for spied on `MongoStorageAdapter.connect()` failing

let's see what else fails.

### CODE REVIEW

**Comment by acinader** [comment]
So here's what I think is going on....

The jasmine upgrade from 3.2 - 3.3 [introduces this change](https://github.com/jasmine/jasmine/blob/master/release_notes/3.2.0.md):
```
Report unhandled rejections as globalErrors.

Merges #1521 from @johnjbarton
```

This is a good thing that will help us prepare for:

```
(node:7741) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 1): Error: woops
(node:7741) DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code. */
```

@stage88 figured out a solution for one set of failing tests with 0eff786

I am now struggling with this test:

https://github.com/parse-community/parse-server/blob/019cf0a98646b8c8c6d194528fc5cd33fb1f35b7/spec/index.spec.js#L60-L83

Which is failing with this uncaught rejection:

```Error: connect ECONNREFUSED 127.0.0.1:43605```

I'm having trouble figuring out where this uncaught rejection is occuring....

**Comment by dplewis** [comment]
https://github.com/parse-community/parse-server/issues/5500#issuecomment-482896110 Check out my comment

**Comment by acinader** [comment]
failing node 11:

``` 1) server fails if database is unreachable
  - Unhandled promise rejection: Error: connect ECONNREFUSED 127.0.0.1:43605
  - Failed: There were open connections to the server left after the test finished
```

**Comment by stage88** [comment]
I should be able to have another look at this test tonight (Sydney, AU - time). 

Sorry if i understood it incorrectly, I think that error ```- Unhandled promise rejection: Error: connect ECONNREFUSED 127.0.0.1:43605``` is expected in that test given there is nothing on ```uri: 'mongodb://fake:fake@localhost:43605/drew3'```.

I can update the test instead to try ```request.catch``` instead of ```request.then```...

**Comment by acinader** [comment]
@stage88 I rolled back #5506 jasmine upgrade and merged #5496 

I think that there is a promise reject leak or two preventing upgrade for jasmine and node 11.

If you can help track those down, would be great.  Thanks for the winston upgrade.  so Helpful!

**Comment by stage88** [comment]
No worries :) I'll have another go at that test tonight.

**Comment by stage88** [comment]
Captured stack trace of where the error originates:

image

**Comment by stage88** [comment]
@acinader check out this commit https://github.com/stage88/parse-server/commit/f131f26fa75c1bac4112e674e7e9f23a9d9aa2ac

All tests are passing, except ```Parse.Query Aggregate testing cannot group by date field (excluding createdAt and updatedAt)```.

**Comment by acinader** [comment]
Nice.  Much improved.  There are a handful of tests failing:

https://travis-ci.org/parse-community/parse-server/jobs/520004000

fails with:

1) ParseServerRESTController should handle a get request with full serverURL mount path
  - Unhandled promise rejection: ParseError: 107 schema class name does not revalidate


https://travis-ci.org/parse-community/parse-server/jobs/520004001

fails with:

1) Server Url Checks handleShutdown, close connection
  - Unhandled promise rejection: Error: Connection pool of the database object has been destroyed.
  - Uncaught exception: TypeError: message.split is not a function
2) Postgres database init options should fail to create server if schema databaseOptions does not exist
  - Unhandled promise rejection: error: no schema has been selected to create in (line 487)


https://travis-ci.org/parse-community/parse-server/jobs/520004002

fails with:

1) SchemaController will resolve class creation races appropriately
  - Unhandled promise rejection: ParseError: 103 Class NewClass already exists.


More broadly, our tests are pretty flakey:

1. https://github.com/parse-community/parse-server/blob/master/spec/helper.js is hard to follow and could use some comments, possibly switch to async / await in a few places to aid readability?

2. we use `setTimeout` ALOT to wait random amounts of time in hopes that async acitivites complete, which is both flakey and makes the tests take much longer than they would if we knew when async activity completed.

I attempted to get some conversation going on the topic here:

https://community.parseplatform.org/t/async-aftersave-notification-on-completion/65


On the tests for my professional projects, I have gotten rid of all the `setTimeouts` by exposing an event emitter that fires when after save hooks complete.

Its rudimentrary for now, but it allows my cloud code to run without blocking clients.  

I am not sure what 'un observed' async code is running in all the places we use `setTimeout` in our tests, but trying to incrementally pick them off by adding an observable/event on completion could help to get our tests more reliable and faster.

And while I am dreaming, I think that one of the biggest services we could do to the parse commun  would be to help make testing easy and well understood for the users of parse, so a blog post or two about how we ended making the tests faster and more reliable....

**Comment by stage88** [comment]
Ah I see, there are are more :) happy to help when I can. 

With the commit I referenced above, I am really not across everything in Parse, I feel uneasy making changes knowing it could negatively impact a lot of Parse users. I would love to spend some more time reviewing that code myself, or maybe get a bit of guidance from maintainers.

Also I agree, there is work that could be done on improving tests, I think many of these chores could be opened up to community, seeking help through "help wanted", etc

**Comment by dplewis** [comment]
Closing via https://github.com/parse-community/parse-server/pull/5573
